### PR TITLE
Support prefix wildcards in gradle --test selectors

### LINKED ISSUE #3121: Conditional tests with Gradle by backward match is failed (5.3.0 or later)
**Which version of Kotest are you using**
- Kotest: 5.3.0 or later
- Kotlin: 1.6.10
- Gradle: 7.4.2

Hi, I use version 5.1.0 of kotest. I have updated my version of kotest to 5.3.0.
`./gradlew test --tests '*Spec'` is failed in version 5.3.0 of kotest.
I browsed https://kotest.io/docs/framework/conditional/conditional-tests-with-gradle.html.
The page says I can filter with `gradle test --tests '*IntegrationTest'`, but failed to run on 5.3.0.

```
$./gradlew test --tests '*Spec'
Starting a Gradle Daemon (subsequent builds will be faster)

> Task :test

NumberSpec > NoSuchElementException FAILED
    java.util.NoSuchElementException at _Strings.kt:71

NumberSpec > executionError FAILED
    java.util.NoSuchElementException at _Strings.kt:71
        Caused by: java.util.NoSuchElementException at _Strings.kt:71

2 tests completed, 2 failed

> Task :test FAILED

FAILURE: Build failed with an exception.
```

Is this a bug?

Thanks.

## Status of Different Versions
Assuming you define somepackage.NumberSpec.

||5.1.0|5.3.0|
|:-----: | :-----: | :-----: |
|backward match <br/> (\*Spec)| ✅ | ❌|
|forward match <br/> (somepackage.\*) |✅|✅|

### Files
- build.gradle
```
plugins {
    id 'org.jetbrains.kotlin.jvm' version '1.6.10'
}

group = 'org.example'
version = '1.0-SNAPSHOT'

repositories {
    mavenCentral()
}

dependencies {
    testImplementation 'io.kotest:kotest-runner-junit5:5.3.0'
}

test {
    useJUnitPlatform()
}

compileTestKotlin {
    kotlinOptions.jvmTarget = '1.8'
}

compileTestKotlin {
    kotlinOptions.jvmTarget = '1.8'
}
```

- Test Target Program File(Number.kt)
```
data class Number(val number: Int) {
    fun addNumber(n: Int): Int {
        return number + n
    }
}
```

- Test Program File (NumberSpec.kt)
```
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe

class NumberSpec : FunSpec({
    test("test") {
        val number = Number(1)

        val actual = number.addNumber(3)
        actual.shouldBe(4)
    }
})
```

### PR DESCRIPTION
**Pull request by sksamuel**
Fixes #3121

### CODE REVIEW

File: kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt [hunk 1]
```diff
@@ -18,41 +18,58 @@ class GradleClassMethodRegexTestFilter(private val patterns: List<String>) : Tes
...
    companion object {
 
       // if the regex starts with a lower case character, then we assume it is in the format package.Class.testpath
       // otherwise, we assume it is in the format Class.testpath
       // the .testpath is always optional, and at least Class or package must be specified
+      // additionally, patterns can start with a * in which case the pattern matches suffixes
       fun parse(pattern: String): GradleTestPattern {
+         require(pattern.isNotBlank())
+
+         val prefixWildcard = pattern.startsWith("*") || pattern.startsWith("*.")
```
Comment (LeoColman) [inline kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt]: If it starts with *. it definitely starts with *, right?
```suggestion
         val prefixWildcard = pattern.startsWith("*")
```

Comment (sksamuel) [reply kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt]: Yes but  I need rid of the . too so we don't end up with the package name being .something

Comment (LeoColman) [reply kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt]: What I mean is
If a pattern start with `*.`, it will always start with `*`.
The second statement won't ever be true if the first one isn't true

Comment (sksamuel) [reply kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt]: Ah yes :)

File: kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt [hunk 2]
```diff
@@ -18,41 +18,58 @@ class GradleClassMethodRegexTestFilter(private val patterns: List<String>) : Tes
...
+         }
+
          is Descriptor.SpecDescriptor -> when {
+            pck != null && classname != null && prefixWildcard -> descriptor.kclass.qualifiedName?.contains("$pck.$classname") ?: false
             pck != null && classname != null -> descriptor.kclass.qualifiedName == "$pck.$classname"
+            pck != null && prefixWildcard -> descriptor.kclass.qualifiedName?.contains(pck) ?: true
             pck != null -> descriptor.kclass.qualifiedName?.startsWith(pck) ?: true
+            classname != null && prefixWildcard -> descriptor.kclass.simpleName?.contains(classname) ?: false
             classname != null -> descriptor.kclass.simpleName == classname
             else -> true
```
Comment (LeoColman) [inline kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt:36]:
```suggestion
         is Descriptor.SpecDescriptor -> when {
            pck == null && classname == null -> true
            pck == null && prefixWildcard -> descriptor.kclass.simpleName?.contains(classname) ?: false
            pck == null -> descriptor.kclass.simpleName == classname
            prefixWildcard -> descriptor.kclass.simpleName?.contains(classname) ?: false
            else -> descriptor.kclass.simpleName == classname
```

I was thinking of maybe inverting the when conditions, would lead to less code although I don't think it's much cleaner hahah

I didn't check if they're the same thing either

File: kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt [hunk 3]
```diff
@@ -18,41 +18,58 @@ class GradleClassMethodRegexTestFilter(private val patterns: List<String>) : Tes
...
          }
       }
    }
 }
 
-data class GradleTestPattern(val pckage: String?, val classname: String?, val path: List<String>) {
+data class GradleTestPattern(
+   val suffixWildcard: Boolean,
```
Comment (LeoColman) [inline kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt]:
```suggestion
   val prefixWildcard: Boolean,
```

File: kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt [hunk 4]
```diff
@@ -18,41 +18,58 @@ class GradleClassMethodRegexTestFilter(private val patterns: List<String>) : Tes
...
 
       // if the regex starts with a lower case character, then we assume it is in the format package.Class.testpath
       // otherwise, we assume it is in the format Class.testpath
       // the .testpath is always optional, and at least Class or package must be specified
+      // additionally, patterns can start with a * in which case the pattern matches suffixes
       fun parse(pattern: String): GradleTestPattern {
+         require(pattern.isNotBlank())
+
+         val prefixWildcard = pattern.startsWith("*")
+         val pattern2 = pattern.removePrefix("*.").removePrefix("*")
```
Comment (LeoColman) [inline kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt]:
```suggestion
         val pattern2 = pattern.removePrefix("*").removePrefix(".")
```

Comment (LeoColman) [reply kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt]: Bit of code golfing lol

File: kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt [hunk 5]
```diff
@@ -18,41 +18,58 @@ class GradleClassMethodRegexTestFilter(private val patterns: List<String>) : Tes
...
 
-         val tokens = pattern.split('.')
+         val tokens = pattern2.split('.')
          val classIndex = tokens.indexOfFirst { it.first().isUpperCase() }
```
Comment (LeoColman) [inline kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt:61]: Not part of the PR, but I'm wondering

It's not possible in the JVM to have a lowercased class, so this necessarily finds something, right?

File: kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt [hunk 6]
```diff
@@ -18,41 +18,58 @@ class GradleClassMethodRegexTestFilter(private val patterns: List<String>) : Tes
...
          // if class is not specified, then we assume the entire string is a package
-         if (classIndex == -1) return GradleTestPattern(pattern, null, emptyList())
+         if (classIndex == -1) return GradleTestPattern(prefixWildcard, pattern2, null, null)
 
          // if the class is the first part, then no package is specified
          val pck = if (classIndex == 0) null else tokens.take(classIndex).joinToString(".")
 
          val pathParts = tokens.drop(classIndex + 1)
-         val path = if (pathParts.isEmpty()) emptyList() else pathParts.joinToString(".").split(Descriptor.TestDelimiter)
+         val path = if (pathParts.isEmpty()) null else pathParts.joinToString(".")
```
Comment (LeoColman) [inline kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt:70]: Isn't this just `pattern2`?

in line 60 we split it, but it seems we are rejoining it here without ever making changes to it.

File: kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt (hunk 5 above)
Comment (sksamuel) [reply kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt:61]: you can name classes what you want.

File: kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt (hunk 6 above)
Comment (sksamuel) [reply kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt:70]: pathPaths is tokens drop 1

File: kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt (hunk 2 above)
Comment (sksamuel) [reply kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt:36]: _I didn't check if they're the same thing either_
well this works so I suggest we leave it ;)

**Review by LeoColman — APPROVED** [review]
